### PR TITLE
nlp torch migration

### PR DESCRIPTION
**Pull request by ymyjl**
add nlp_fast_explore and nlp_migrate_experiences

### CODE REVIEW

File: docs/guides/model_convert/NLP_fast_explore.md [hunk 1]
```diff
@@ -0,0 +1,1711 @@
...
+
+如果输出下面的内容，则说明 PaddlePaddle 安装成功。
+
+```plain
+PaddlePaddle is installed successfully! Let's start deep learning with PaddlePaddle now.2.2.0
+```
+
+
+
+1. 安装 Pytorch
```
Comment (ZHUI) [inline docs/guides/model_convert/NLP_fast_explore.md]:
```suggestion
2. 安装 Pytorch
```

File: docs/guides/model_convert/NLP_fast_explore.md [hunk 2]
```diff
@@ -0,0 +1,1711 @@
...
+import torch
+print(torch.__version__)
+# 如果安装的是 gpu 版本，可以按照下面的命令确认 torch 是否安装成功
+# 期望输出为 tensor([1.], device='cuda:0')
+print(torch.Tensor([1.0]).cuda())
+```
+
+
+
+\3. 安装差异核验工具 reprod_log
```
Comment (ZHUI) [inline docs/guides/model_convert/NLP_fast_explore.md]:
```suggestion
3. 安装差异核验工具 reprod_log
```

File: docs/guides/model_convert/NLP_fast_explore.md [hunk 3]
```diff
@@ -0,0 +1,1711 @@
...
+```bash
+pip3 install reprod_log --force-reinstall
+```
+
+
+
+### 2.2 准备数据
+
+- 了解该模型输入输出格式：Bert 句子分类任务为例，模型输入为句子组，也就是维度为`[batch_size, sequence_length]`的 tensor，而其中具体的值单词需要通过 tokenizer 转换为数字 id， 类型为 int64，label 为`[batch_size, ]`的 label，类型为`int64`。
+- 准备伪输入数据（fake input data）以及伪标签（fake label）：通过运行生成伪数据的参考代码：https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/torch_migration/pipeline/fake_data/gen_fake_data.py，生成和模型输入 shape、type 等保持一致的伪数据，并保存在本地，用于后续模型前反向对齐时同时作为两个模型的输入。这样的方式能够使保证模型对齐不会受到输入数据的影响，有助于将模型结构对齐和数据对齐解耦，更为方便地排查问题。伪数据可以通过如下代码生成。
```
Comment (ZHUI) [inline docs/guides/model_convert/NLP_fast_explore.md]:
```suggestion
- 准备伪输入数据（fake input data）以及伪标签（fake label）：通过运行生成伪数据的参考代码：https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/torch_migration/pipeline/fake_data/gen_fake_data.py ，生成和模型输入 shape、type 等保持一致的伪数据，并保存在本地，用于后续模型前反向对齐时同时作为两个模型的输入。这样的方式能够使保证模型对齐不会受到输入数据的影响，有助于将模型结构对齐和数据对齐解耦，更为方便地排查问题。伪数据可以通过如下代码生成。
```

File: docs/guides/model_convert/NLP_fast_explore.md [hunk 4]
```diff
@@ -0,0 +1,1711 @@
...
+
+
+
+### 2.3 分析并运行参考代码
+
+需在特定设备(CPU/GPU)上，利用少量伪数据，跑通参考代码的预测过程(前向)以及至少 2 轮(iteration)迭代过程，用于生成和迁移代码进行对比的结果。
+
+PyTorch 的实现：
+
+[bert_torch](https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/torch_migration/pipeline/Step5/bert_torch)
```
Comment (ZHUI) [inline docs/guides/model_convert/NLP_fast_explore.md]:
```suggestion
PyTorch 的实现：
[bert_torch](https://github.com/PaddlePaddle/PaddleNLP/tree/develop/examples/torch_migration/pipeline/Step5/bert_torch)
```

File: docs/guides/model_convert/NLP_fast_explore.md [hunk 5]
```diff
@@ -0,0 +1,1711 @@
...
+
+对应转换后的 PaddlePaddle 实现: [paddle_bert](https://github.com/PaddlePaddle/PaddleNLP/blob/develop/examples/torch_migration/pipeline/models/pd_bert.py)。
+
+> 飞桨 PaddleNLP 官方实现的代码请参考：[paddlenlp bert/modeling.py](https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/transformers/bert/modeling.py)
+
+**【源码分析】**
+
+分析 Bert 网络代码，主要分为以下几个模块：
+
+- [BertEmbedding](https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/transformers/bert/modeling.py#:~:text=class BertEmbeddings)：定义了 BERT 输入的分布式表示。输入嵌入是 Token 嵌入、Segment 嵌入和 Position 嵌入的总和。
```
Comment (ZHUI) [inline docs/guides/model_convert/NLP_fast_explore.md]:
```suggestion
- [BertEmbedding](https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/transformers/bert/modeling.py#:~:text=class%20BertEmbeddings)：定义了 BERT 输入的分布式表示。输入嵌入是 Token 嵌入、Segment 嵌入和 Position 嵌入的总和。
```

File: docs/guides/model_convert/NLP_fast_explore.md [hunk 6]
```diff
@@ -0,0 +1,1711 @@
...
+
+![](./pictures/embedding.png)
+
+图 BertEmbedding 模块
+
+- [EncoderLayer](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/nn/layer/transformer.py#:~:text=class TransformerEncoderLayer)：继承自 `torch.nn.Layer`，是 Bert 网络中基本模块，由 MultiHeadAttention、FeedForward 组成。后者由 LayerNorm，Dropout，Linear 层和激活函数构成。
```
Comment (ZHUI) [inline docs/guides/model_convert/NLP_fast_explore.md]:
```suggestion
- [EncoderLayer](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/nn/layer/transformer.py#:~:text=class%20TransformerEncoderLayer)：继承自 `torch.nn.Layer`，是 Bert 网络中基本模块，由 MultiHeadAttention、FeedForward 组成。后者由 LayerNorm，Dropout，Linear 层和激活函数构成。
```

Comment (ZHUI) [reply docs/guides/model_convert/NLP_fast_explore.md]: 后面的链接，同理。空格改成 %20

File: docs/guides/model_convert/NLP_fast_explore.md [hunk 7]
```diff
@@ -0,0 +1,1711 @@
...
+
+![](./pictures/encoder.png)
```
Comment (ZHUI) [inline docs/guides/model_convert/NLP_fast_explore.md]: 图片的大小宽度，需要调整。
类似这里，设置宽度
https://github.com/PaddlePaddle/PaddleNLP/blob/develop/model_zoo/ernie-1.0/pretraining_introduction.md?plain=1#L288-L290

**Review by ZHUI — COMMENTED** [review]
主要一些链接问题需要修改。图片格式问题等

File: docs/guides/model_convert/NLP_migration_experiences.md [hunk 8]
```diff
@@ -0,0 +1,1176 @@
...
+3. **小数据集数据读取对齐**：数据读取对齐为了验证数据加载、数据预处理、数据增强与原始代码一致。为了快速验证数据读取对齐，建议准备一个小数据集（训练集和验证集各 8~16 张图像即可，压缩后数据大小建议在`20M`以内）。
+4. **评估指标对齐**：评估指标是模型精度的度量。在计算机视觉中，不同任务使用的评估指标有所不同（比如，在图像分类任务中，常用的指标是 Top-1 准确率与 Top-5 准确率；在图像语义分割任务中，常用的指标是 mIOU）为了检验迁移后的模型能否达到原模型的精度指标，需要保证使用的评估指标与原始代码一致以便对照。飞桨提供了一系列 Metric 计算类，而 PyTorch 中目前可以通过组合的方式实现或者调用第三方的 API。
+5. **损失函数对齐**：损失函数是训练模型时的优化目标，使用的损失函数会影响模型的精度。在模型迁移时，需要保证迁移后模型训练时使用的损失函数与原始代码中使用的损失函数一致，以便二者对照。飞桨与 PyTorch 均提供了常用的损失函数。
+6. **模型训练超参对齐**：模型的训练超参包括学习率、优化器、正则化策略等。这些超参数指定了模型训练过程中网络参数的更新方式，训练超参数的设置会影响到模型的收敛速度及收敛精度。同样地，在模型迁移时，需要保证迁移前后模型使用的训练超参数一致，以便对照二者的收敛情况。飞桨中的 optimizer 有`paddle.optimizer`等一系列实现，PyTorch 中则有`torch.optim`等一系列实现。完成超参对齐后，可以使用反向梯度对齐统一验证该模块的正确性。
+7. **反向梯度对齐**：在完成前向对齐的基础上，还需进行反向梯度对齐。反向梯度对齐的目的是确保迁移后的模型反向传播以及权重更新的行为与原始模型一致，同时也是对上一步*模型训练超参对齐*的验证。具体的检验方法是通过两次（或以上）迭代训练进行检查，若迁移前后的模型第二轮训练的 loss 一致，则可以认为二者反向已对齐。
+8. **训练集数据读取对齐：**相同的神经网络使用不同的数据训练和测试得到的结果往往会存在差异。因此，为了能复现原始代码的精度，需要保证使用的数据完全相同，包括数据集的版本、使用的数据预处理方法和流程、使用的数据增强方式等。
```
Comment (ZHUI) [inline docs/guides/model_convert/NLP_migration_experiences.md]: 这个 markdown 展示有点问题，

**Comment by Ligoml** [comment]
https://github.com/PaddlePaddle/docs/pull/5545 已合入
